File: 3448.count-substrings-divisible-by-last-digit.py
#
# @lc app=leetcode id=3448 lang=python3
#
# [3448] Count Substrings Divisible By Last Digit
#

# @lc code=start
class Solution:
    def countSubstrings(self, s: str) -> int:
        count = 0
        n = len(s)
        # Iterate over all possible end indices of substrings
        for end in range(n):
            last_digit = int(s[end])
            if last_digit == 0:
                continue  # Skip if last digit is zero, as no divisibility possible
            # Initialize current number starting from s[end] as a single digit number
            num = 0
            # Extend the substring backwards checking divisibility by last_digit
            for start in range(end, -1, -1):
                num = num + int(s[start]) * (10 ** (end - start))  # Update number ending at 'end' starting from 'start'
                if num % last_digit == 0:
                    count += 1  # If divisible, increment count
        return count  # Return total count of such substrings
# @lc code=end

File: 3448.count-substrings-divisible-by-last-digit.cpp
#
# @lc app=leetcode id=3448 lang=cpp
#
# [3448] Count Substrings Divisible By Last Digit
#

# @lc code=start
class Solution {
public:
    long long countSubstrings(string s) {
        long long count = 0;
        int n = s.size();
        for (int i = 0; i < n; ++i) {
            int lastDigit = s[i] - '0';
            if (lastDigit == 0) continue; // Skip zero as divisor
            long long num = 0;
            for (int j = i; j < n; ++j) {
                num = num * 10 + (s[j] - '0');
                if (num % lastDigit == 0) {
                    ++count;
                }
            }
        }
        return count;
    }
};
# @lc code=end